presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSubsetNodeSetSupplier.java

```diff
@@ -0,0 +1,296 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.execution.scheduler.NodeSelectionHashStrategy.CONSISTENT_HASHING;
+import static com.facebook.presto.metadata.InternalNode.NodeStatus.ALIVE;
+import static com.facebook.presto.spi.NodeState.ACTIVE;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.lang.Math.ceil;
+import static java.lang.String.format;
+
+/**
+ * This WIP class is responsible for assigning a subset of the cluster nodes
+ * to a requesting query.
+ * The size of the subset is currently static, but it can by defined by configuration
+ * or can be driven by fairness algorithm depending on query metadata.
+ */
+public class FixedSubsetNodeSetSupplier
+        implements NodeSetSupplier, Closeable
+{
+    private static final Logger log = Logger.get(FixedSubsetNodeSetSupplier.class);
+    private final PriorityQueue<NodeSetAcquireRequest> pendingRequests;
+    private final InternalNodeManager nodeManager;
+    private final NetworkLocationCache networkLocationCache;
+    // TODO: Use Set<InternalNode>
+    private final Map<QueryId, List<InternalNode>> perQueryNodeAssignments;
+
+    private final ScheduledFuture<?> scheduledFuture;
+
+    @Inject
+    public FixedSubsetNodeSetSupplier(
+            InternalNodeManager nodeManager,
+            NetworkTopology networkTopology,
+            @ForNodeScheduler ScheduledExecutorService scheduledExecutorService)
+    {
+        this.nodeManager = nodeManager;
+        this.networkLocationCache = new NetworkLocationCache(networkTopology);
+        this.perQueryNodeAssignments = new HashMap<>();
+        this.pendingRequests = new PriorityQueue<>();
+        // start loop to check pending node requests every second
+        scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(this::fulfillPendingRequests, 0, 100, TimeUnit.MILLISECONDS);
+    }
+
+    private void fulfillPendingRequests()
+    {
+        if (pendingRequests.size() == 0) {
+            return;
+        }
+
+        log.info("Trying to fulfill nodeset request for=%s", pendingRequests.peek().queryId);
+        List<InternalNode> allocatedNodes = perQueryNodeAssignments.values().stream()
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        List<InternalNode> currentActiveWorkerNodesInCluster = nodeManager.getNodes(ACTIVE).stream()
+                .filter(node -> !node.isResourceManager() && !node.isCoordinator() && !node.isCatalogServer())
+                .collect(toImmutableList());
+
+        List<InternalNode> avaialableNodes = currentActiveWorkerNodesInCluster.stream()
+                .filter(node -> !allocatedNodes.contains(node))
```

Comment:
I wonder how likely these lookups to impact test results? With thousands of nodes in a cluster that may be taking a while. Might be safer to do a `Set`.

Reply:
Good point. Will update